azure/services/managedclusters/managedclusters.go

```diff
@@ -183,23 +183,64 @@ func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
-	} else if !azure.ResourceNotFound(err) {
+	}
+
+	isCreate := azure.ResourceNotFound(err)
+	if isCreate {
+		err = s.Client.CreateOrUpdate(ctx, managedClusterSpec.ResourceGroupName, managedClusterSpec.Name, managedCluster)
+		if err != nil {
+			return fmt.Errorf("failed to create managed cluster, %w", err)
+		}
+	} else {
 		ps := *existingMC.ManagedClusterProperties.ProvisioningState
 		if ps != "Canceled" && ps != "Failed" && ps != "Succeeded" {
-			klog.V(2).Infof("Unable to update existing managed cluster in non terminal state.  Managed cluster must be in one of the following provisioning states: canceled, failed, or succeeded")
+			s.Scope.V(2).Info("Unable to update existing managed cluster in non terminal state.  Managed cluster must be in one of the following provisioning states: canceled, failed, or succeeded")
 			return nil
 		}
-	}
 
-	err = s.Client.CreateOrUpdate(ctx, managedClusterSpec.ResourceGroupName, managedClusterSpec.Name, properties)
-	if err != nil {
-		return fmt.Errorf("failed to create or update managed cluster, %w", err)
+		// Normalize properties for the existing managed cluster, so that we
+		// check only those fields that were specified in the initial
+		// CreateOrUpdate request. Without comparing to normalized properties,
+		// we would always get a difference in desired and existing, which
+		// would result in sending unnecessary Azure API requests.
+		existingMCPropertiesNormalized := &containerservice.ManagedClusterProperties{
+			NodeResourceGroup:       existingMC.ManagedClusterProperties.NodeResourceGroup,
+			DNSPrefix:               existingMC.ManagedClusterProperties.DNSPrefix,
+			KubernetesVersion:       existingMC.ManagedClusterProperties.KubernetesVersion,
+			LinuxProfile:            existingMC.ManagedClusterProperties.LinuxProfile,
+			ServicePrincipalProfile: existingMC.ManagedClusterProperties.ServicePrincipalProfile,
+			AgentPoolProfiles:       &[]containerservice.ManagedClusterAgentPoolProfile{},
+			NetworkProfile: &containerservice.NetworkProfileType{
+				NetworkPlugin:   existingMC.ManagedClusterProperties.NetworkProfile.NetworkPlugin,
+				LoadBalancerSku: existingMC.ManagedClusterProperties.NetworkProfile.LoadBalancerSku,
+				NetworkPolicy:   existingMC.ManagedClusterProperties.NetworkProfile.NetworkPolicy,
+			},
+		}
+
+		if managedClusterSpec.ServiceCIDR != "" {
+			// We will check ServiceCidr and DNSServiceIP only in the case they
+			// were set in the initial request, so basically under the same
+			// conditions like above when setting values in managedCluster.
+			if managedClusterSpec.DNSServiceIP == nil {
+				existingMCPropertiesNormalized.NetworkProfile.ServiceCidr = existingMC.ManagedClusterProperties.NetworkProfile.ServiceCidr
+			}
+			existingMCPropertiesNormalized.NetworkProfile.DNSServiceIP = existingMC.ManagedClusterProperties.NetworkProfile.DNSServiceIP
+		}
+
+		diff := cmp.Diff(managedCluster.ManagedClusterProperties, existingMCPropertiesNormalized)
```

Comment:
This was already done for agent pool `CreateOrUpdate` requests, so I did the same thing here. One of the consequences of this approach is that any manual change on the Azure AKS resource will remain unaffected, until some part of the CR changes so the new `CreateOrUpdate` is sent.

OTOH there are not many properties that are configurable in AKS after the cluster has been created (especially the properties that are currently supported in the CAPZ managed cluster types), so this might be fine for now.

Reply:
> any manual change on the Azure AKS resource will remain unaffected, until some part of the CR changes so the new CreateOrUpdate is sent.

what would the expected behavior be in your mind? that's what the normalization was intended to cover. we only support a subset of properties, so we only set those and diff against those. once CAPZ supports any given property, we should be diffing/applying it.

If we don't support a property, how would we send something other than what was done out of band? If we do support it, we should be covering the diff here, converging on 100% support.

Reply:
I missed this comment, sorry about that.

I was maybe overthinking it here a bit, you're right, if some property is supported by CAPZ, it should be included in the diff, otherwise no need to check it. Another limiting factor here is if Azure supports the modification of the property or not.

Out of all properties that are set here for managed cluster creation https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/02ca6c383c4c459efb79c8d3673f332f212f4046/azure/services/managedclusters/managedclusters.go#L123, it seems to me that only the Kubernetes version can be updated. All other properties are either hard-coded in CAPZ, or they are used only during creation and Azure does not allow to modify them (e.g. like Linux profile), so I don't think that there is a point in checking them.

Here are the properties that can be updated via CLI https://docs.microsoft.com/en-us/cli/azure/aks?view=azure-cli-latest#az_aks_update, now I guess that you can modify a bit more via API directly, but I'm not sure how much more, I couldn't find any docs that explain it, since Azure API for managed clusters does not have separate Create and Update requests, but only a single `CreateOrUpdate` request.

To be more concrete, out of all managed cluster properties that are set [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/02ca6c383c4c459efb79c8d3673f332f212f4046/azure/services/managedclusters/managedclusters.go#L123), I don't think that these can be modified:
- Location
- ManagedClusterProperties.NodeResourceGroup
- ManagedClusterProperties.DNSPrefix
- ManagedClusterProperties.LinuxProfile
- ManagedClusterProperties.NetworkProfile
  - NetworkPlugin
  - LoadBalancerSKU
  - NetworkPolicy
  - PodCidr
  - DNSServiceIP

These are hard-coded in CAPZ for now:
- Identity
- ManagedClusterProperties.ServicePrincipalProfile

AgentPoolProfiles are handled by agentpools service, so no need to check them in managed cluster update as well.

Which then leaves only ManagedClusterProperties.KubernetesVersion as the only property here which can be updated.

Am I making any sense here? :smile:

Reply:
I said I would reply later so here I am 🙂 I think I said pretty much everything that mattered in slack, however. A lot of properties in the AKS API are, as you noticed, immutable. So it definitely makes sense to remove them from the diff and block them earlier, e.g. in webhooks. Up to you whether you want to tackle that in this PR :)

Reply:
> Up to you whether you want to tackle validation in this PR, or stick to fixing the errors

Finally got some time to finish this. I have added update validations to the AzureManagedControlPlane webhook that are returning an error when an immutable field is changed.